### ISSUE TITLE
Shadowing Section

### ISSUE THREAD

**Issue by SpiralOSS** [fsprojects/fsharp-cheatsheet#31]
Maybe shadowing deserves an individual section? you used `open type` above for another reason, but for me a key thing that trips people up is how DU cases shadow others and/or type names. I always used to open a DU's module, but really `open type` on the DU brings less stuff in.

Not sure the "AuotOpen would sidestep thi"s is going to be grokked by someon scanning a cheatsheet...

_Originally posted by @bartelink in https://github.com/fsprojects/fsharp-cheatsheet/pull/30#discussion_r1390366431_

**Comment by IanWitham** [comment]
`[<RequireQualifiedAccess>]` is another way to avoid shadowing problems. Perhaps this could be added to the code organisation section?

**Comment by IanWitham** [comment]
I think the `AutoOpen` section could just include something like:

> Use [<AutoOpen>] with caution as functions with identical names from different modules will silently shadow each other, causing the most recently imported definition to be used instead of the one you might expect.

**Comment by bartelink** [comment]
Yes, something like that (though it needs to convey the sheer pain a random unavoidable reordering can cause if things are badly enough named in a big enough codebase!)

**Comment by SpiralOSS** [comment]
> Yes, something like that (though it needs to convey the sheer pain a random unavoidable reordering can cause if things are badly enough named in a big enough codebase!)

But also: cheat sheet. We don't have to convey everything.

**Comment by bartelink** [comment]
I guess I was trying to nudge it in the direction of:
> Use `AutoOpen` and/or ignore the `RequiredQualifiedAccess` best practice with caution as functions (and union cases) with identical names from different modules will silently shadow each other, causing compile errors if the `open` order changes.
But I agree that omitting stuff is good too. Main thing for me is that AutoOpen is not mentioned without a caveat, and that RQA gets a mention

**Comment by IanWitham** [comment]
Yeah my RQA example is quite verbose.
